[legacy] invalid passing of bool contract method arguments

## TL;DR
With the `legacy` target for web3-0.2.x, `bool` arguments in contract methods are passed through a `.toString()` in the generated TS class, which results in the argument always being set to `true` on the blockchain side. 

Did not test with other targets.

## Minimal reproduction
### Contract
```solidity
contract TestContract {
    bool public field;
    event FieldSet(bool value);

    function setField(bool _field) public {
        field = _field;
        emit FieldSet(field);
    }
}
```

### Test
```ts
import { artifacts, contract, web3 } from './truffle-globals';
import { TestContract } from '@contracts/TestContract';
import { expect } from 'chai';
import { it } from 'mocha';

contract('TestContract', accounts => {
    let testInstance: TestContract;
    before(async () => {
        const truffleDeployedContract = await artifacts
            .require(TestContract.name)
            .new({ from: accounts[0] });
        testInstance = await TestContract.createAndValidate(
            web3,
            truffleDeployedContract.address,
        );
    });

    it('tests setting the field', async () => {
        expect(await testInstance.field).to.eq(false);
        await testInstance.setFieldTx(true).send({ from: accounts[0] });
        expect(await testInstance.field).to.eq(true);
    });

    it('tests unsetting the field', async () => {
        expect(await testInstance.field).to.eq(true);
        await testInstance.setFieldTx(false).send({ from: accounts[0] });
        expect(await testInstance.field).to.eq(false);
    });
});

```

### Test output
Most important detail: during the second test the emitted event says `value: true`, even though the method was called with `false`:
```
  Contract: TestContract
    ✓ tests setting the field (68ms)
    1) tests unsetting the field

    Events emitted during test:
    ---------------------------

    FieldSet(value: true)

    ---------------------------


  1 passing (264ms)
  1 failing

  1) Contract: TestContract
       tests unsetting the field:

      AssertionError: expected true to equal false
      + expected - actual

      -true
      +false
      
      at Object.<anonymous> (test/asdf.spec.ts:27:45)
      at Generator.next (<anonymous>)
      at fulfilled (test/asdf.spec.ts:4:58)
      at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:182:7)
```

### Culprit
The `_field.toString()` argument in `TestContract.ts`.
When the `.toString()` conversion is removed manually, the test passes as expected.
```ts
  public setFieldTx(
    _field: boolean
  ): TC.DeferredTransactionWrapper<TC.ITxParams> {
    return new TC.DeferredTransactionWrapper<TC.ITxParams>(this, "setField", [
      _field.toString()
    ]);
  }
```

### env
`typechain 0.3.8`
`web3 0.20.7`
`ganache-cli 6.1.8`
`truffle 4.1.13`

--- a/lib/targets/legacy/generation.ts
+++ b/lib/targets/legacy/generation.ts
@@ -146,9 +146,10 @@ function codeGenForParams(param: AbiParameter, index: number): string {
 }
 
 function codeGenForArgs(param: AbiParameter, index: number): string {
-  const isArray = param.type instanceof ArrayType;
   const paramName = param.name || `arg${index}`;
-  return isArray ? `${paramName}.map(val => val.toString())` : `${paramName}.toString()`;
+  if (param.type instanceof ArrayType) return `${paramName}.map(val => val.toString())`;
+  if (param.type instanceof BooleanType) return paramName;
+  return `${paramName}.toString()`;
 }
 
 function codeGenForOutputTypeList(output: Array<AbiParameter>): string {

--- a/lib/targets/legacy/generation.ts
+++ b/lib/targets/legacy/generation.ts
@@ -147,7 +147,10 @@ function codeGenForParams(param: AbiParameter, index: number): string {
 
 function codeGenForArgs(param: AbiParameter, index: number): string {
   const paramName = param.name || `arg${index}`;
-  if (param.type instanceof ArrayType) return `${paramName}.map(val => val.toString())`;
+  if (param.type instanceof ArrayType) {
+    const elemParam = { name: `${paramName}Elem`, type: param.type.itemType };
+    return `${paramName}.map(${elemParam.name} => ${codeGenForArgs(elemParam, 0)})`;
+  }
   if (param.type instanceof BooleanType) return paramName;
   return `${paramName}.toString()`;
 }

--- a/scripts/test.sh
+++ b/scripts/test.sh
@@ -1,6 +1,5 @@
 #!/usr/bin/env bash
 set -e
-set -x
 cd "$(dirname "$0")"
 cd ..
 

--- a/scripts/test.sh
+++ b/scripts/test.sh
@@ -1,5 +1,6 @@
 #!/usr/bin/env bash
 set -e
+set -x
 cd "$(dirname "$0")"
 cd ..
 

--- a/test/integration/contracts/DumbContract.sol
+++ b/test/integration/contracts/DumbContract.sol
@@ -67,6 +67,11 @@ contract DumbContract {
     return boolParam;
   }
 
+  function callWithBooleanArray(bool[] boolArrayParam) public pure returns (bool[]) {
+    boolArrayParam = boolArrayParam;
+    return boolArrayParam;
+  }
+
   function testAddress(address a) pure public returns (address) {
     return a;
   }

--- a/test/integration/targets/legacy/DumbContract.spec.ts
+++ b/test/integration/targets/legacy/DumbContract.spec.ts
@@ -195,6 +195,12 @@ describe("DumbContract", () => {
     expect(result).to.eq(byteString);
   });
 
+  it("should correctly pass both boolean values in args", async () => {
+    const dumbContract = await DumbContract.createAndValidate(web3, contractAddress);
+    expect(await dumbContract.callWithBoolean(true)).to.be.deep.eq(true);
+    expect(await dumbContract.callWithBoolean(false)).to.be.deep.eq(false);
+  });
+
   describe("estimateGas", () => {
     it("should work", async () => {
       const dumbContract = await DumbContract.createAndValidate(web3, contractAddress);

--- a/test/integration/targets/legacy/DumbContract.spec.ts
+++ b/test/integration/targets/legacy/DumbContract.spec.ts
@@ -197,8 +197,16 @@ describe("DumbContract", () => {
 
   it("should correctly pass both boolean values in args", async () => {
     const dumbContract = await DumbContract.createAndValidate(web3, contractAddress);
+
     expect(await dumbContract.callWithBoolean(true)).to.be.deep.eq(true);
     expect(await dumbContract.callWithBoolean(false)).to.be.deep.eq(false);
+  });
+
+  it("should correctly pass a boolean array argument", async () => {
+    const dumbContract = await DumbContract.createAndValidate(web3, contractAddress);
+    const array = [true, false];
+
+    expect(await dumbContract.callWithBooleanArray(array)).to.be.deep.eq(array);
   });
 
   describe("estimateGas", () => {
@@ -207,7 +215,7 @@ describe("DumbContract", () => {
 
       const estimatedGas = await dumbContract.countupTx(1).estimateGas({ from: accounts[0] });
 
-      expect(estimatedGas).to.be.eq(82683);
+      expect(estimatedGas).to.be.eq(82705);
     });
   });
 });

